File: leetcode/0085-maximalRectangle.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0)
            return 0;
        int area = 0;
        int[] line = new int[matrix[0].length];
        boolean[] isProcessed = new boolean[matrix[0].length];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                if (matrix[i][j] == '1') {
                    line[j]++;
                    isProcessed[j] = false;
                } else {
                    line[j] = 0;
                    isProcessed[j] = true;
                } 
            }
            int len = matrix[i].length;
            for (int j = 0; j < len; j++) {
                if (isProcessed[j])
                    continue;
                int leftMove = 1, rightMove = 1;
                while (j+rightMove  < len && line[j] <= line[j+rightMove]) {
                    if (line[j] == line[j+rightMove])
                        isProcessed[j+rightMove] = true;
                    rightMove++;
                }
                while (j-leftMove >= 0 && line[j] <= line[j-leftMove]) {
                    leftMove++;
                }
                int curArea = line[j] * (rightMove + leftMove - 1);
                area = Math.max(curArea, area);
            }
        }
        return area;
    }
}

File: leetcode/0085-maximalRectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or len(matrix) == 0 or len(matrix[0]) == 0:
            return 0
        area = 0
        line = [0 for i in range(len(matrix[0]))]
        is_processed = [False for i in range(len(matrix[0]))]
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                if matrix[i][j] == "1":
                    line[j] += 1
                    is_processed[j] = False
                else:
                    line[j] = 0
                    is_processed[j] = True
            length = len(matrix[i])
            for j in range(length):
                if is_processed[j]:
                    continue
                left_move, right_move = 1, 1
                while j+right_move < length and line[j] <= line[j+right_move]:
                    if line[j] == line[j+right_move]:
                        is_processed[j+right_move] = True
                    right_move += 1
                while j-left_move >= 0 and line[j] <= line[j-left_move]:
                    left_move += 1
                cur_area = line[j] * (right_move + left_move - 1)
                area = max(cur_area, area)
        return area
